bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -232,104 +224,99 @@ static inline void jucx_call_callback(jobject callback, jobject jucx_request,
-                                      (native_ptr)request);
-        struct jucx_context *ctx = (struct jucx_context *)request;
-        ucs_recursive_spin_lock(&ctx->lock);
-        if (ctx->status == UCS_INPROGRESS) {
-            // request not completed yet, install user callback
-            if (callback != NULL) {
-                ctx->callback = env->NewGlobalRef(callback);
-            }
-            ctx->jucx_request = env->NewGlobalRef(jucx_request);
-        } else {
-            // request was completed whether by progress in other thread or inside
-            // ucp_tag_recv_nb function call.
-            set_jucx_request_completed(env, jucx_request, ctx, ctx->status);
-            if (callback != NULL) {
-                jucx_call_callback(callback, jucx_request, ctx->status);
-            }
-            jucx_context_reset(ctx);
-            ucp_request_free(request);
-        }
-        ucs_recursive_spin_unlock(&ctx->lock);
-    } else {
-        jmethodID empty_constructor = env->GetMethodID(jucx_request_cls, "<init>", "()V");
-        jucx_request = env->NewObject(jucx_request_cls, empty_constructor);
-        set_jucx_request_completed(env, jucx_request, NULL, UCS_PTR_RAW_STATUS(request));
-        if (UCS_PTR_IS_ERR(request)) {
-            JNU_ThrowExceptionByStatus(env, UCS_PTR_STATUS(request));
-            if (callback != NULL) {
-                call_on_error(callback, UCS_PTR_STATUS(request));
-            }
-        } else if (callback != NULL) {
-            call_on_success(callback, jucx_request);
-        }
-    }
-    return jucx_request;
+    jobject jucx_request = reinterpret_cast<jobject>(user_data);
+
+    jucx_request_update_sender_tag(env, jucx_request, info->sender_tag);
+    jucx_request_update_recv_length(env, jucx_request, info->length);
+    jucx_request_callback(request, status, user_data);
 }
 
-jobject process_completed_stream_recv(size_t length, jobject callback)
+void stream_recv_callback(void *request, ucs_status_t status, size_t length,
+                          void *user_data)
 {
     JNIEnv *env = get_jni_env();
-    jobject jucx_request = env->NewObject(jucx_request_cls, jucx_request_constructor, NULL);
-    env->SetObjectField(jucx_request, native_id_field, NULL);
-    env->SetLongField(jucx_request, recv_size_field, length);
-    jucx_request_update_status(env, jucx_request, UCS_OK);
+    jobject jucx_request = reinterpret_cast<jobject>(user_data);
+    jucx_request_update_recv_length(env, jucx_request, length);
+
+    jucx_request_callback(request, status, user_data);
+}
+
+jobject jucx_request_allocate(JNIEnv *env, const jobject callback,
+                              ucp_request_param_t *param, jint memory_type)
+{
+    jobject jucx_request = env->NewObject(jucx_request_cls, jucx_request_constructor);
+
+    param->op_attr_mask = UCP_OP_ATTR_FIELD_USER_DATA  |
+                          UCP_OP_ATTR_FIELD_CALLBACK   |
```

Comment:
too many spaces before |

Reply:
Ok, thought to align by the longest word